kotlinx-coroutines-core/jvm/src/debug/AgentPremain.kt

```diff
@@ -0,0 +1,65 @@
+/*
+ * Copyright 2016-2020 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license.
+ */
+
+package kotlinx.coroutines.debug
+
+import kotlinx.coroutines.debug.internal.DebugProbesImpl
+import sun.misc.*
+import java.lang.instrument.*
+import java.lang.instrument.ClassFileTransformer
+import java.security.*
+
+@Suppress("unused")
+internal object AgentPremain {
+
+    private val enableCreationStackTraces =
+        System.getProperty("kotlinx.coroutines.debug.enable.creation.stack.trace")?.toBoolean()
+            ?: DebugProbesImpl.enableCreationStackTraces
+
+    @JvmStatic
+    public fun premain(args: String?, instrumentation: Instrumentation) {
+        instrumentation.addTransformer(DebugProbesTransformer)
+        DebugProbesImpl.enableCreationStackTraces = enableCreationStackTraces
+        DebugProbesImpl.install()
+        installSignalHandler()
+    }
+
+    internal object DebugProbesTransformer : ClassFileTransformer {
+        override fun transform(
+            loader: ClassLoader,
+            className: String,
+            classBeingRedefined: Class<*>?,
+            protectionDomain: ProtectionDomain,
+            classfileBuffer: ByteArray?
+        ): ByteArray? {
+            val name = className.replace("/", ".")
+            if (name != "kotlin.coroutines.jvm.internal.DebugProbesKt") {
+               return null
+            }
+            /*
+             * DebugProbesKt.bin contains `kotlin.coroutines.jvm.internal.DebugProbesKt` class
+             * with method bodies that delegate all calls directly to their counterparts in
+             * kotlinx.coroutines.debug.DebugProbesImpl. This is done to avoid classfile patching
+             * on the fly (-> get rid of ASM dependency)
+             */
```

Comment:
The `DebugProbesKt.bin` file should be generated during build of the project (as opposed to being committed as a byte blob).

Reply:
Do we really want to? It's a lot of burden to maintain, more Gradle machinery, some risk of accidentally including this classfile in the JAR (not as a resource), using unofficial flag `-Xallow-kotlin-package` etc. And benefits are really unclear

Reply:
Two reasons:
* Transparency: Just an opaque `.bin` file looks spooky.
* Maintenance: What if we need to change it in the future (move Impl to different package or rename, etc) and the secret on how to produce this file is lost.

Reply:
Let's compromise on the middle ground :)